src/pytorch/messages/passMessagesCUDA.cu

```diff
@@ -0,0 +1,123 @@
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/cuda/CUDAStream.h>
+#include <torch/extension.h>
+
+#include "common/accessor.cuh"
+#include "common/atomicAdd.cuh"
+
+using c10::cuda::CUDAStreamGuard;
+using c10::cuda::getCurrentCUDAStream;
+using torch::autograd::AutogradContext;
+using torch::autograd::Function;
+using torch::autograd::tensor_list;
+using torch::kInt32;
+using torch::Tensor;
+using torch::TensorOptions;
+
+template <typename scalar_t> __global__ void kernel_forward(
+    const Accessor<int32_t, 2> neighbors,
+    const Accessor<scalar_t, 2> messages,
+    Accessor<scalar_t, 2> new_states
+) {
+    const int32_t i_neig = blockIdx.x;
+    const int32_t i_dir = blockIdx.y;
+    const int32_t i_atom = neighbors[i_dir][i_neig];
+    if (i_atom < 0) return;
+
+    const int32_t i_feat = threadIdx.x;
+    atomicAdd(&new_states[i_atom][i_feat], messages[i_neig][i_feat]);
```

Comment:
You can eliminate the limitations on number of features by just rewriting this as a loop.

```c++
for (int32_t i_feat = threadIdx.x; i_feat < num_features; i_feat += blockDim.x)
    atomicAdd(&new_states[i_atom][i_feat], messages[i_neig][i_feat]);
```

Reply:
Apart from solving a non-existing problem, this would make the memory access not coalesced and reduce speed.

Reply:
It would have no effect on speed at all.  If the number of features happens to satisfy your current requirement, the behavior would be identical to what it currently does.  The `atomicAdd()` would be executed once by every thread with `i_feat` equal to `threadIdx.x`.  The only change would be if the number doesn't satisfy your current requirements, either because it's not a multiple of 32 or it's more than 1024.  In that case it would produce correct behavior, unlike the current code.  So there's no downside at all.

Reply:
It will reduce the number of thread by the number of features. The reduced parallelism would result into reduced speed.

Reply:
I'm not suggesting any change to the number of threads.  The only thing I'm suggesting is wrapping the `atomicAdd()` in a loop as shown above.  If `num_features` happens to match your current restrictions, nothing will change.  Every thread will still call it exactly once.